### PR TITLE
Fixes issues with models that use guarded & purgeable attributes

### LINKED ISSUE #5313: BackendAuth::register usage
I use the example code:
```
$user = BackendAuth::register([
    'name' => 'Some User',
    'login' => 'someuser',
    'email' => 'some@website.tld',
    'password' => 'changeme',
    'password_confirmation' => 'changeme'
]);
```
Location: https://octobercms.com/docs/backend/users#backend-auth-facade

I got the following error:
error

### PR DESCRIPTION
**Pull request by LukeTowers**
Fixes octobercms/october#5313. Related: https://github.com/octobercms/library/commit/b779df0174454fea203c10d7582465f9ddaf22fb

### CODE REVIEW

File: src/Database/Concerns/GuardsAttributes.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        // JSON columns are tricksy, we only guard base level columns though
+        if (strpos($key, '->') !== false) {
+            $key = Str::before($key, '->');
+        }
+
+        return in_array($key, $guarded);
+    }
+}
```
Comment (LukeTowers) [inline src/Database/Concerns/GuardsAttributes.php]:
```suggestion
}

```

File: tests/Database/ModelTest.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        $this->assertTrue($model->isGuarded('data->key'));
+    }
+}
+
+class TestModel extends Model
+{
+    protected $guarded = ['id', 'ID', 'data'];
+
+    public $table = 'test_model';
+}
```
Comment (LukeTowers) [inline tests/Database/ModelTest.php]:
```suggestion
}

```

File: src/Database/Concerns/GuardsAttributes.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        // Normalize the variables for comparison
+        $key = trim(strtolower($key));
+        $guarded = array_map(function ($column) {
+            return trim(strtolower($column));
+        }, $this->getGuarded());
+
+        // Nothing's guarded so just return early
+        if (empty($guarded) || $guarded === ['*']) {
+            return false;
+        }
```
Comment (bennothommo) [inline src/Database/Concerns/GuardsAttributes.php]: @LukeTowers this is a micro-optimization for sure, but I think it's best to check that there is no guard first before running any array modifications. I can safely assume that `isGuarded` will be called several times throughout the course of a data-heavy action.

```suggestion
        $guarded = $this->getGuarded();
        
        // Nothing's guarded so just return early
        if (empty($guarded) || $guarded === ['*']) {
            return false;
        }
        
        // Normalize the variables for comparison
        $key = trim(strtolower($key));
        $guarded = array_map(function ($column) {
            return trim(strtolower($column));
        }, $guarded);
```

File: tests/Database/ModelTest.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    public function testFalse()
+    {
+        $model = new TestModel();
+
+        // Test base guarded property
+        $this->assertTrue($model->isGuarded('data'));
+
+        // Test variations on casing
+        $this->assertTrue($model->isGuarded('DATA'));
+
```
Comment (bennothommo) [inline tests/Database/ModelTest.php:35]: @LukeTowers additional test for the reverse (lower case key, upper-case guarded value)
```suggestion
        $this->assertTrue($module->isGuarded('name'));

```

File: tests/Database/ModelTest.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        // Test JSON columns
+        $this->assertTrue($model->isGuarded('data->key'));
+    }
+}
+
+class TestModel extends Model
+{
+    protected $guarded = ['id', 'ID', 'data'];
```
Comment (bennothommo) [inline tests/Database/ModelTest.php]: For above:
```suggestion
    protected $guarded = ['id', 'ID', 'NAME', 'data'];
```

**Comment by LukeTowers** [comment]
@bennothommo are you able to test this against the original issue? Also, do you have any recommendations for setting up DB level unit testing? I wanted to use unit tests that were closer to Laravel's for this feature but it was such a hassle and I wasn't able to get it working well enough.

**Comment by RomainMazB** [comment]
The fix is working for me :clap:

File: src/Database/Concerns/GuardsAttributes.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+     * and allows fields that don't exist in the database to be filled if they aren't specified as "guarded", under
+     * the pretense that they are handled in a special manner - ie. in the `beforeSave` event.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    public function isGuarded($key)
+    {
+        $guarded = $this->getGuarded();
+
```
Comment (mjauvin) [inline src/Database/Concerns/GuardsAttributes.php]:
```suggestion

```

File: src/Database/Concerns/GuardsAttributes.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    public function isGuarded($key)
+    {
+        $guarded = $this->getGuarded();
+        
+        // Nothing's guarded so just return early
+        if (empty($guarded) || $guarded === ['*']) {
+            return false;
+        }
+
```
Comment (mjauvin) [inline src/Database/Concerns/GuardsAttributes.php]:
```suggestion

```

**Review by mjauvin — COMMENTED** [review]
empty lines with spaces only.

**Comment by mjauvin** [comment]
@LukeTowers I tested this PR and it's working as advertized!

**Comment by LukeTowers** [comment]
Great, thanks @mjauvin! I'm just waiting on getting the tests to actually use the DB like they do for Laravel's version of the guarded attribute tests before merging this, pretty busy with a bunch of other projects right now.

**Comment by bennothommo** [comment]
I'm convinced this fix works as expected :) Nice one, @LukeTowers!

**Comment by LukeTowers** [comment]
Thanks for the assistance on this one @bennothommo! Are we able to get this particular fix backported to 1.0.470 or is it more trouble than it's worth?

**Comment by bennothommo** [comment]
@LukeTowers might as well - we could port the fix, not the tests perhaps, then it's only a two-file edit.

**Pull request merged**